### PR TITLE
feat: file purger

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

Implement file purger that delete SST files marked as deleted, for example, when SST file is compacted.

Finally we won't see disk usage grows after table compaction :sweat_smile:

Some design considerations:
- File purger reuses the `LocalScheduler`, as we will need throttle when invoking API to delete files on OSS;
   - Purge request's key is region id + sst file name to avoid potential confilict.
- Add `delete_sst` method to `AccessLayer` trait;
- `FileHandle` holds a reference to `AccessLayer` and `FilePurger`. When `FileHandleInner` is dropped, it schedules a purge request to `FilePurger` if it has been marked deleted;
    - Instead of `ObjectStore`, `FileHandle` holds `AccessLayerRef`, because SST files in different regions may use different `ObjectStore`, since `ObjectStore` is per bucket when backed by OSS.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#930

### CODE REVIEW

File: src/storage/src/error.rs [hunk 1]
```diff
@@ -416,6 +416,12 @@ pub enum Error {
         source: JoinError,
         backtrace: Backtrace,
     },
+
+    #[snafu(display("Failed to delete SST file, source: {:?}", source))]
```
Comment (evenyag) [inline src/storage/src/error.rs:420]:
```suggestion
    #[snafu(display("Failed to delete SST file, source: {}", source))]
```

File: src/storage/src/sst.rs [hunk 2]
```diff
@@ -202,21 +221,58 @@ impl FileHandle {
...
+            };
+            match self.file_purger.schedule(request) {
+                Ok(res) => {
+                    info!(
+                        "Scheduled SST purge task, region: {}, name: {}, res: {}",
+                        self.meta.region_id, self.meta.file_name, res
+                    );
+                }
+                Err(e) => {
+                    error!(e; "Failed to schedule SST purge task, region: {}, name: {}",self.meta.region_id, self.meta.file_name);
```
Comment (evenyag) [inline src/storage/src/sst.rs:245]:
```suggestion
                    error!(e; "Failed to schedule SST purge task, region: {}, name: {}", self.meta.region_id, self.meta.file_name);
```

File: src/storage/src/file_purger.rs [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+impl Handler for FilePurgeHandler {
+    type Request = FilePurgeRequest;
+
+    async fn handle_request(
+        &self,
+        req: Self::Request,
+        token: BoxedRateLimitToken,
+        finish_notifier: Arc<Notify>,
+    ) -> crate::error::Result<()> {
+        req.sst_layer.delete_sst(&req.file_name).await?;
```
Comment (evenyag) [inline src/storage/src/file_purger.rs:51]: Add a log if delete_sst returns error?
```suggestion
        req.sst_layer.delete_sst(&req.file_name).await.map_err(|e| error!(); e)?;
```

**Review by waynexia — APPROVED** [review]
Mostly looks good to me

File: src/storage/src/engine.rs [hunk 4]
```diff
@@ -234,6 +237,13 @@ impl<S: LogStore> EngineInner<S> {
     ) -> Self {
         let job_pool = Arc::new(JobPoolImpl {});
         let flush_scheduler = Arc::new(FlushSchedulerImpl::new(job_pool));
+
+        let file_purger = Arc::new(LocalScheduler::new(
+            SchedulerConfig {
+                max_inflight_task: 32, //we allow more file purge tasks to run concurrently.
```
Comment (killme2008) [inline src/storage/src/engine.rs:243]: Why not let the user able to configure it?

File: src/storage/src/file_purger.rs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        ));
+        let (handle, path, _layer) =
+            create_sst_file(object_store.clone(), sst_file_name, scheduler.clone()).await;
+
+        {
+            // mark file as deleted and drop the handle, we expect the file is deleted.
+            handle.mark_deleted();
+            drop(handle);
+        }
+        scheduler.stop().await.unwrap();
```
Comment (killme2008) [inline src/storage/src/file_purger.rs:220]: Maybe we can add a `flush` function for `scheduler` to wait all in-flight tasks finished.

File: src/storage/src/sst.rs [hunk 6]
```diff
@@ -202,21 +221,58 @@ impl FileHandle {
...
     meta: FileMeta,
     compacting: AtomicBool,
+    deleted: AtomicBool,
+    sst_layer: AccessLayerRef,
+    file_purger: FilePurgerRef,
+}
+
+impl Drop for FileHandleInner {
+    fn drop(&mut self) {
+        if self.deleted.load(Ordering::Relaxed) {
```
Comment (killme2008) [inline src/storage/src/sst.rs:231]: Do we need to check `compacting` too? Is it possible that a compacting file is marked to be deleted?

Comment (v0y4g3r) [reply src/storage/src/sst.rs:231]: Currently no, files are only marked as deleted iff compaction succeeds.

Comment (v0y4g3r) [reply src/storage/src/sst.rs:231]: > Do we need to check `compacting` too? Is it possible that a compacting file is marked to be deleted?

Currently no, files are only marked as deleted iff compaction succeeds.

File: src/storage/src/file_purger.rs (hunk 5 above)
Comment (v0y4g3r) [reply src/storage/src/file_purger.rs]: > Maybe we can add a `flush` function for `scheduler` to wait all in-flight tasks finished.

I [added an parameter `await_termination: bool` to `Scheduler::stop` method](https://github.com/GreptimeTeam/greptimedb/pull/1030/commits/d5fe3ec8a2577d6cdf9e53097e94a526d3787382#diff-7d1a7794d06be67a5a6cca11de718380af37d859b115c2db7139f996d5087a03R133). If set to true, the scheduler will wait until all queued requests are processed, PTAL.

File: src/datanode/src/datanode.rs [hunk 7]
```diff
@@ -113,13 +113,16 @@ pub struct CompactionConfig {
     pub max_inflight_task: usize,
```
Comment (killme2008) [inline src/datanode/src/datanode.rs:113]: I think these parameters should be in the plural. For example: `max_inflight_tasks`, `max_files_in_level0` and` max_purge_tasks`.

File: src/storage/src/engine.rs (hunk 4 above)
Comment (killme2008) [reply src/storage/src/engine.rs:243]: It's still in hard coded.

File: src/storage/src/scheduler.rs [hunk 8]
```diff
@@ -90,6 +98,8 @@ pub struct LocalScheduler<R: Request> {
     task_notifier: Arc<Notify>,
     /// Join handle of spawned request handling loop.
     join_handle: Mutex<Option<JoinHandle<()>>>,
+    /// State of scheduler.
+    state: Arc<AtomicI32>,
```
Comment (killme2008) [inline src/storage/src/scheduler.rs:102]: Maybe `u8` is enough.

File: src/storage/src/scheduler.rs [hunk 9]
```diff
@@ -186,36 +212,51 @@ where
 {
     /// Runs scheduled requests dispatch loop.
     pub async fn run(&self) {
-        let task_notifier = self.task_notifier.clone();
         let limiter = self.limiter.clone();
-        loop {
+        while self.state.load(Ordering::Relaxed) == STATE_RUNNING {
```
Comment (killme2008) [inline src/storage/src/scheduler.rs:216]:
```suggestion
        while self.running() {
```

File: src/storage/src/scheduler.rs [hunk 10]
```diff
@@ -186,36 +212,51 @@ where
...
-                    return;
+                    info!("Task scheduler cancelled.");
+                    break;
                 }
             }
         }
+        // For correctness, we need to poll requests from fifo again.
+        if self.state.load(Ordering::Relaxed) == STATE_AWAIT_TERMINATION {
+            info!("Waiting for all pending tasks to finish.");
+            self.poll_and_execute(&limiter).await;
```
Comment (killme2008) [inline src/storage/src/scheduler.rs:231]: Do we need to set the state to be `STATE_STOP` after waiting?

**Review by killme2008 — COMMENTED** [review]
Leave some comments. 
But i would like to implement `flush` in scheduler by adding a special flushing task. After the flusing task is processed,then we can make sure that the previous tasks are all processed too. And this `flush` function can be also used in `await_termination`.

**Comment by v0y4g3r** [comment]
> Leave some comments. But i would like to implement `flush` in scheduler by adding a special flushing task. After the flusing task is processed,then we can make sure that the previous tasks are all processed too. And this `flush` function can be also used in `await_termination`.

That would be a little tricky since `Request` is an associate type of scheduler instead of a trait object. So I wonder in what case should we need this functionality?

**Comment by killme2008** [comment]
> That would be a little tricky since `Request` is an associate type of scheduler instead of a trait object. So I wonder in what case should we need this functionality?

For example, while closing the region when failover, it requires flushing all inflight tasks associated with the region.
P.S.  `max_file_in_level0` should be renamed to `max_files_in_level0`. The `max_files` looks like the maximum size of a file.

**Comment by v0y4g3r** [comment]
> > That would be a little tricky since `Request` is an associate type of scheduler instead of a trait object. So I wonder in what case should we need this functionality?
> 
> For example, while closing the region when failover, it requires flushing all inflight tasks associated with the region. P.S. `max_file_in_level0` should be renamed to `max_files_in_level0`. The `max_file` looks like the maximum size of a file.

I think in that case we can return something like `JoinHandle` from `schedule` method and let the caller to decide whether to store the handle and wait for all scheduled task to finish or just cancel them? After all when a region is closing, it does not have to wait for all previously scheduled tasks in which there are tasks scheduled by other regions, it only cares about its own tasks.

**Comment by killme2008** [comment]
> > > That would be a little tricky since `Request` is an associate type of scheduler instead of a trait object. So I wonder in what case should we need this functionality?
> > 
> > 
> > For example, while closing the region when failover, it requires flushing all inflight tasks associated with the region. P.S. `max_file_in_level0` should be renamed to `max_files_in_level0`. The `max_file` looks like the maximum size of a file.
> 
> I think in that case we can return something like `JoinHandle` from `schedule` method and let the caller to decide whether to store the handle and wait for all scheduled task to finish or just cancel them? After all when a region is closing, it does not have to wait for all previously scheduled tasks in which there are tasks scheduled by other regions, it only cares about its own tasks.

Yep, we can leave it later.

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**